File: pa1/Lex.java
import java.io.*;
import java.util.Scanner;

class Lex{
	public static void main (String[] args) throws IOException{
		
	  // check number of command line arguments is at least 2
      if(args.length < 2){
         System.err.println("Usage: java Lex <input file> <output file>");
         System.exit(1);
      }
	  
	  // open files
      BufferedReader reader = new BufferedReader(new FileReader(args[0]));
      PrintWriter out = new PrintWriter(new FileWriter(args[1]));
	  
	  //while loop to count number of lines within in file
	  int linecount = 0;
      while( reader.readLine() != null ){
         linecount++;
	  }
	  reader.close();
	  
	  //Second reader to read lines and add individual lines to array
	  BufferedReader reader2 = new BufferedReader(new FileReader(args[0]));
	  String[] array = new String[linecount];
	  int index_tracker = 0;
	  String line = "";
	  while( (line = reader2.readLine()) != null ){
		 array[index_tracker] = line;
		 index_tracker++;
	  }
	  //List creation and sorting (insertion sort)
	  List list = new List();
	  for(int i = 0; i < array.length; i++){
		  if(i == 0){
			  list.prepend(i);
		  }else{
			  list.moveBack();
			  boolean inserted = false;
			  while(inserted == false){
				  int check = array[i].compareTo(array[list.get()]);
				  if(check <= 0){
					  if(list.cursor == list.front){
						  list.prepend(i);
						  inserted = true;
					  }else{
						  list.movePrev();
				      }
				  }else{
					  if(list.cursor == list.back){
						  list.append(i);
						  inserted = true;
					  }else{
						  list.insertAfter(i);
						  inserted = true;
					  }
				  }
			  }
		  }
	  }
	  
	  list.moveFront();
	  while(list.cursor != null){
		  System.out.println(array[list.get()]);
		  list.moveNext();
	  }
	  
	  // close files
      reader2.close();
      out.close();
	}
}

File: grading_scripts/pa2/Lex.c
#include<stdio.h>
#include<stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "List.h"

int main(int argc, char* argv[]){

   //Check if input is valid. Throws a Usage error if it is not
   if(argc != 3){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(EXIT_FAILURE);
   }
 
   FILE* in;  /* file handle for input */  
   FILE* out; /* file handle for output */
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   
   //Opens file to count number of words for array size
   int length = 0;
   char line[256];
   while(fgets(line,256,in) != NULL){
      length++;
   } 
   fclose(in);
   
   //Re-opens file to read line-by-line and add words to array
   in = fopen(argv[1], "r");
   char *array[length];
   int counter = 0;
   while(fgets(line,256,in) != NULL){
      array[counter] = malloc(strlen(line)+1);
      strcpy(array[counter],line);
      counter++;
   }   
   //List creation and sorting (insertion sort)
   List list = newList();
   for(int i = 0; i < length; i++){
      if(i == 0){
         prepend(list, i);
      }else{
         moveBack(list);
         bool inserted = false;
         while(inserted == false){
            int check = strcmp(array[i],array[get(list)]);
            if(check <= 0){
               if(isFront(list) == 1){
                  prepend(list,i);
                  inserted = true;
               }else{
                  movePrev(list);
               }
            }else{
               if(isBack(list) == 1){
                  append(list,i);
                  inserted = true;
               }else{
                  insertAfter(list,i);
                  inserted = true;
               }
            }
         }
      }
   }  

   //Prints lexicographic sorting of input file to output file
   moveFront(list);
   for(int i = 0; i < length; i++){
      fprintf(out,array[get(list)]);
      moveNext(list);
   }

   //Closes files
   fclose(in);
   fclose(out);

   //Frees memory   
   freeList(&list);
   for(int i = 0; i < length; i++){
      free(array[i]);
   }
}
